### PR TITLE
D-Logic Support and Fault creation to be merged on Master (for everyone) 

### PR DESCRIPTION
**Pull request by AlooshieJ**
There should not be any conflicts, I only added to the code.
This is here for you to make the pull on your master branch.

### CODE REVIEW

File: circuit_utils/circuit_simulator.py [hunk 1]
```diff
@@ -141,8 +143,32 @@ def prompt(self):
...
+                else:
+                    final_inputs.append(input_values[chars]) # this will always be a single D
+
+        #Debug tool to see values
+        #print(input_values, final_inputs)
+        for character, node in zip(final_inputs, self.nodes.input_nodes.values()):
             node.set(nodes.Value(character))
+
+        # asking for faulty node
+        with_fault = input("Do you want a fauly node? (y/n)")
```
Comment (dhodczak) [inline circuit_utils/circuit_simulator.py]: typo on line 167

File: circuit_utils/circuit_simulator.py [hunk 2]
```diff
@@ -157,11 +183,45 @@ def simulate(self):
...
+            if node.name == is_node:
+                print(f"found the node {node.name} = {node.value}")
+                found_node = node
+                break
+
+        if is_node not in self.nodes:
+            print("Not a valid node to change")
+            return False
+        else:
+            fault_value = input(f"which value do you want node {found_node.name} to  be stuck at? (1/0)")
```
Comment (dhodczak) [inline circuit_utils/circuit_simulator.py]: typo on line 200 : two spaces

File: circuit_utils/circuit_simulator.py [hunk 3]
```diff
@@ -157,11 +183,45 @@ def simulate(self):
...
+
+            #else: #this should be an exception?
+            #    pass
+
+
         pass
 
     def detect_fault(self):
         #     TODO: Detect any faults that have propagated to the outputs
-        pass
```
Comment (dhodczak) [inline circuit_utils/circuit_simulator.py:208]: cleanup

File: circuit_utils/circuit_simulator.py [hunk 4]
```diff
@@ -157,11 +183,45 @@ def simulate(self):
...
+        if any(node == 'D' or node == "D'" for node in self.nodes.output_nodes):
+            s = f" Fault detected by propagation of D / D'" \
+                f""
```
Comment (dhodczak) [inline circuit_utils/circuit_simulator.py]: cleanup

File: circuit_utils/nodes.py [hunk 5]
```diff
@@ -10,6 +10,12 @@ def __init__(self, value: str):
             self.value = 0
         elif value == 1 or value == '1':
             self.value = 1
+
+        # Final value for Ds wil be string "D" / "D'"
+        elif value == 'D' or value =='d':
+            self.value = 'D'
```
Comment (dhodczak) [inline circuit_utils/nodes.py]: should be using nodes.Value("D"), not "D"

File: circuit_utils/nodes.py [hunk 6]
```diff
@@ -10,6 +10,12 @@ def __init__(self, value: str):
             self.value = 0
         elif value == 1 or value == '1':
             self.value = 1
+
+        # Final value for Ds wil be string "D" / "D'"
+        elif value == 'D' or value =='d':
```
Comment (dhodczak) [inline circuit_utils/nodes.py]: Do not have to compare to both cases -- can just compare to "D"

File: circuit_utils/circuit_simulator.py [hunk 7]
```diff
@@ -157,11 +183,45 @@ def simulate(self):
...
+            #else: #this should be an exception?
+            #    pass
+
+
         pass
 
     def detect_fault(self):
         #     TODO: Detect any faults that have propagated to the outputs
-        pass
+        if any(node == 'D' or node == "D'" for node in self.nodes.output_nodes):
```
Comment (dhodczak) [inline circuit_utils/circuit_simulator.py]: This method doesn't seem to account for faults in several output nodes

File: circuit_utils/nodes.py [hunk 8]
```diff
@@ -23,6 +29,14 @@ def __eq__(self, other):
         elif self.value == 'U':
             if other == 'U' or other == 'u':
                 return True
+        #added D value here
+        elif self.value == 'D':
+            if other == 'd' or other == 'D':
```
Comment (dhodczak) [inline circuit_utils/nodes.py:34]: Do not have to check both lower and upper--cleanup

**Review by dhodcz2 — CHANGES_REQUESTED** [review]

File: circuit_utils/nodes.py (hunk 5 above)
Comment (AlooshieJ) [reply circuit_utils/nodes.py]: should I add this to the 1 and 0 cases as well?

Comment (dhodczak) [reply circuit_utils/nodes.py]: Yes. Do not set self.value to a character -- set it to a Value(1), Value('D'), etc

**Pull request merged**